lib/src/rules/collection_methods_unrelated_type.dart

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../analyzer.dart';
+import '../util/unrelated_types_visitor.dart';
+
+const _desc = 'Invocation of various collection methods with arguments of '
+    'unrelated types.';
+
+const _details = r'''
+
+**DON'T** invoke `contains` on `Iterable` with an instance of different type
+than the parameter type.
```

Comment:
This looks like it needs to be generalized.

Also, the documentation is long, which looks like a signal to me that having a single lint to cover all of these cases is probably a bad idea.

Reply:
Done